docs/environment_variables.md
# Environment Variables

| Variable Name             | Description                                                                                                                                                                        | Required                       | Default                |
| ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------ | ---------------------- |
| DB_DRIVE                  | Determines which database type to use. `postgres` database is supported.                                                                                              | YES                            | -                      |
| DB_HOST                   | Represents the url or ip address of the database that needs to be connected.                                                                                                       | YES                            | -                      |
| DB_PORT                   | Represents the port of the database that needs to be connected.                                                                                                                    | YES                            | -                      |
| DB_NAME                   | Represents the name of the database that needs to be connected.                                                                                                                    | YES                            | -                      |
| DB_USER                   | Represents the user of the database that needs to be connected.                                                                                                                    | YES                            | -                      |
| DB_PASSWORD               | Represents the password of the database that needs to be connected.                                                                                                                | YES                            | -                      |
| DB_MIGRATION_TABLE_NAME   | Represents the name of the table that will be created to store the migration history.                                                                                              | NO                             | auth_service_migration |
| JWT_ALGORITHM             | Variable is represents the method of encryption used to secure and validate JSON Web Tokens (JWTs)                                                                                 | YES                            | -                      |
| JWT_AUDIENCE              | Variable is identifies the target recipient or audience of the JWT. It signifies the application or service that is expected to accept and handle the token.                       | YES                            | -                      |
| JWT_ISSUER                | Variable is designates the source entity that generated the JWT. It indicates the authorization server or entity that is accountable for creating and digitally signing the token. | YES                            | -                      |
| JWT_SECRET                | Variable is a confidential key known only to the issuer and recipient, utilized for signing and verifying the JWT. It guarantees the token's authenticity and integrity.           | YES                            | -                      |
| JWT_EXPIRES_IN            | Variable is defines the time duration, in seconds, during which a JSON Web Token (JWT) remains valid after its creation.                                                           | NO                             | 3600                   |
| SWAGGER_ENABLED           | Variable is used to enable or disable Swagger documentation for an API.                                                                                                            | NO                             | false                  |
| CORS_ALLOWED_ORIGINS      | Allowed origins for cors configuration.                                                                                                                                            | NO                             | \*                     |
| USER_DEFAULT_ROLE         | Represents the role that the user will be added by default during user registration.                                                                                               | NO                             | -                      |
| EMAIL_SERVICE             | Specifies the email service to be used. Only `smtp` and `aws` parameters are accepted and it's mandatory.                                                                          | YES                            | -                      |
| AWS_SES_REGION            | Represents the region that AWS SES Service used.                                                                                                                                   | Required (if AWS is selected)  | -                      |
| AWS_SES_ACCESS_KEY_ID     | Access key id for AWS SES service.                                                                                                                                                 | Required (if AWS is selected)  | -                      |
| AWS_SES_SECRET_ACCESS_KEY | Secret access key for AWS SES service.                                                                                                                                             | Required (if AWS is selected)  | -                      |
| SMTP_HOST                 | Represents the hostname of the SMTP server.                                                                                                                                        | Required (if SMTP is selected) | -                      |
| SMTP_USER                 | Represents the username for SMTP authentication.                                                                                                                                   | Required (if SMTP is selected) | -                      |
| SMTP_PASSWORD             | Represents the password for SMTP authentication.                                                                                                                                   | Required (if SMTP is selected) | -                      |
| OTP_EXPIRES_IN            | Represents the time duration, in seconds, during which an OTP (One-Time Password) remains valid after its creation. authentication.                                                | NO                             | 180                    |
| EMAIL_FROM                | Represents the sender email address used for outgoing emails.                                                                                                                      | YES                            | -                      |
| LOGIN_OTP_EMAIL_SUBJECT   | Represents the subject of the email sent for login otp requests.                                                                                                                   | NO                             | Login Otp              |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.14.4](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.14.3...v1.14.4) (2024-04-26)


### Bug Fixes

* fix e2e test for remove jest mock ([3ab1953](https://github.com/BrewInteractive/authentication-service-nestjs/commit/3ab1953faad68591e9835088a53d3f4454b53179))
* fix for app module test ([3de7e09](https://github.com/BrewInteractive/authentication-service-nestjs/commit/3de7e091d573c5b9fc30508d6478cca0d398ae53))
* fixed password validation. ([daac595](https://github.com/BrewInteractive/authentication-service-nestjs/commit/daac595b8a791291fb799c04f72e251c7da8e029))

### [1.14.3](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.14.2...v1.14.3) (2024-01-23)

### [1.14.2](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.14.1...v1.14.2) (2024-01-23)

### [1.14.1](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.14.0...v1.14.1) (2024-01-22)


### Bug Fixes

* fix subject issue for MailService.sendEmail ([9727b68](https://github.com/BrewInteractive/authentication-service-nestjs/commit/9727b687f0e4c15e667236ffb11027ef77a665f9))

## [1.14.0](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.13.1...v1.14.0) (2024-01-05)


### Features

* Add basePath configuration to support API versioning ([8865d8c](https://github.com/BrewInteractive/authentication-service-nestjs/commit/8865d8caa7af5cc3c5f3576eb8b8d99662360c62))

### [1.13.1](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.13.0...v1.13.1) (2024-01-04)


### Bug Fixes

* Update user-reset-password-request entity and migrations ([fec87e8](https://github.com/BrewInteractive/authentication-service-nestjs/commit/fec87e8bc5e64c56d8337228cc9830b146c46100))

## [1.13.0](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.12.0...v1.13.0) (2024-01-04)


### Features

* **TS-209:** add send reset password email request ([f0f8ac0](https://github.com/BrewInteractive/authentication-service-nestjs/commit/f0f8ac0c80979242a8bca03be85a0f92df715646))


### Bug Fixes

* dynamic email service environment value for e2e tests ([911c722](https://github.com/BrewInteractive/authentication-service-nestjs/commit/911c722ed90cef4a01a9675a87e5565a4289fab4))
* format project ([3c87cb9](https://github.com/BrewInteractive/authentication-service-nestjs/commit/3c87cb908ebf15f0fe21d20efbf9dc9599a17777))
* mock config() in reset-password.module.spec.ts ([126deec](https://github.com/BrewInteractive/authentication-service-nestjs/commit/126deecd53aba5a3f6265bb8accf342c889f63d9))
* remove unnecessary type for sonar issue ([4148eca](https://github.com/BrewInteractive/authentication-service-nestjs/commit/4148ecab9f211fa8d1b00398dbe8a68bd883d3a3))
* **TS-209:** fix for app module test ([4306812](https://github.com/BrewInteractive/authentication-service-nestjs/commit/4306812db91666f6477929661844827b9921f151))
* **TS-252:** fix app module test ([38fb66b](https://github.com/BrewInteractive/authentication-service-nestjs/commit/38fb66bad275b403ba32d4d8cf733c873d13cd07))
* **TS-252:** fix for app and reset password test ([f9ddcb6](https://github.com/BrewInteractive/authentication-service-nestjs/commit/f9ddcb6cc254549a02679c68b6d1da9c9da18880))

## [1.12.0](https://github.com/BrewInteractive/authentication-service-nestjs/compare/v1.11.1...v1.12.0) (2023-12-25)


### Features

* add user module ref in PluginModule ([28b25d3](https://github.com/BrewInteractive/authentication-service-nestjs/commit/28b25d313f78047ec58ecf9821cc0054fde71e25))
* plugin module integrated ([8fd1d38](https://github.com/BrewInteractive/authentication-service-nestjs/commit/8fd1d3884a850d0de49d41f498cddca87986d1a3))
* **TS-187:** create refresh-token module ([920a27c](https://github.com/BrewInteractive/authentication-service-nestjs/commit/920a27cd9916bfb3226536f7c124e52783a1c40d))
* **TS-204:** add template module tests. ([d8e8845](https://github.com/BrewInteractive/authentication-service-nestjs/commit/d8e88452d13370c9bac996d1651c4859462dc709))
* **TS-204:** add template plugin to read me. ([615ec30](https://github.com/BrewInteractive/authentication-service-nestjs/commit/615ec30d9e5ad1ebd494011a049f825fd0bf5cdc))
* **TS-204:** add test case for html. ([b730b46](https://github.com/BrewInteractive/authentication-service-nestjs/commit/b730b46a3f0a2116d5f3de70a6e877ba29ed1355))
* **TS-204:** injectData now only compiles mjml template and getResetPasswordEmailTemplate now only return html. ([1a76495](https://github.com/BrewInteractive/authentication-service-nestjs/commit/1a76495c2d62c80a591092fadf8e288e33819498))
* **TS-204:** made service editable via plugin and inject data now accepts html. ([1ee8612](https://github.com/BrewInteractive/authentication-service-nestjs/commit/1ee861222e15fe766f65e3a9eca207f97c011747))
* **TS-204:** template service and getResetPasswordEmail method. ([1375c78](https://github.com/BrewInteractive/authentication-service-nestjs/commit/1375c78643ec2657a1d344b1d07417caa35b9ff5))
* **TS-241:** implemented createRefreshTokenAsync func. ([8aa4158](https://github.com/BrewInteractive/authentication-service-nestjs/commit/8aa415808ea8b6247eefd39f8bd6962b05ca94bf))
* **TS-253:** update refresh token method ([3236a52](https://github.com/BrewInteractive/authentication-service-nestjs/commit/3236a527959f6d0985dc91e832b47ebddcd170d6))


### Bug Fixes

* fix for migrations ([b48fff2](https://github.com/BrewInteractive/authentication-service-nestjs/commit/b48fff28b9baddc86ca2f33cab75e64a49403251))
* **TS-204:** add property to ts-config esModuleInterop ([27887f0](https://github.com/BrewInteractive/authentication-service-nestjs/commit/27887f0c80af1fc577ffaf5263e70e522b7b0d77))

src/config/configuration.ts
import { description, name, version } from "../../package.json";

export default () => ({
  // Application info
  version,
  name,
  description,

  // Server configurations
  environment: process.env.ENVIRONMENT || "dev",
  port: process.env.PORT || 3000,
  basePath: process.env.BASE_PATH || "/",
  corsAllowedOrigins: process.env.CORS_ALLOWED_ORIGINS || "*",
  swaggerEnabled: process.env.SWAGGER_ENABLED === "true" || false,

  // Authentication & authorization
  apiKey: process.env.API_KEY,
  userDefaultRole: process.env.USER_DEFAULT_ROLE,
  passwordRegex: new RegExp(
    process.env.PASSWORD_REGEX || "(?=.*[A-Z])(?=.*[a-z]).*"
  ),
  jwt: {
    algorithm: process.env.JWT_ALGORITHM,
    audience: process.env.JWT_AUDIENCE,
    issuer: process.env.JWT_ISSUER,
    secret: process.env.JWT_SECRET,
    expiresIn: process.env.JWT_EXPIRES_IN || 3600,
  },
  refreshToken: {
    expiresIn: parseInt(process.env.REFRESH_TOKEN_EXPIRES_IN) || 86400,
  },

  otp: {
    expiresIn: parseInt(process.env.OTP_EXPIRES_IN) || 180,
  },

  // AWS configurations
  aws: {
    sesRegion: process.env.AWS_SES_REGION,
    sesAccessKey: process.env.AWS_SES_ACCESS_KEY_ID,
    sesSecretKey: process.env.AWS_SES_SECRET_ACCESS_KEY || "us-east-1",
  },

  // Email Service
  emailService: process.env.EMAIL_SERVICE,

  // Smtp configurations
  smtp: {
    host: process.env.SMTP_HOST,
    auth: {
      user: process.env.SMTP_USER,
      pass: process.env.SMTP_PASSWORD,
    },
  },
});
